Updating Functions ARM API specs for ANT 82 updates

mathewc opened the pull request:
Spec updates for the set of Functions ARM API changes that were made in ANT 82. Notes on the changes:

- fixed up some description string grammar
- Added new Types: **KeyInfo**, **HostKeys**
- Added the following properties to existing **FunctionEnvelope** type: `test_data_href`, `invoke_url_template`, `language`
- Added the following new APIs:
  - POST `api/sites/{name}[/slots/{slot}]/functions/{functionName}/listkeys`
  - PUT `api/sites/{name}[/slots/{slot}]/functions/{functionName}/keys/{keyName}`
  - DELETE `api/sites/{name}[/slots/{slot}]/functions/{functionName}/keys/{keyName}`
  - POST `api/sites/{name}[/slots/{slot}]/host/default/listkeys`
  - PUT `api/sites/{name}[/slots/{slot}]/host/default/{functionkeys|systemkeys}/{keyName}`
  - DELETE `api/sites/{name}[/slots/{slot}]/host/default/{functionkeys|systemkeys}/{keyName}`
  - POST `api/sites/{name}[/slots/{slot}]/host/default/sync`
  - POST `api/sites/{name}[/slots/{slot}]/host/default/listsyncstatus`

mathewc committed:
Functions ARM API Swagger updates

AutorestCI commented:
# Automation for azure-sdk-for-ruby

AutorestCI commented:
# Automation for azure-sdk-for-python

AutorestCI commented:
# Automation for azure-sdk-for-go

AutorestCI commented:
# Automation for azure-sdk-for-java

azuresdkci commented:
Can one of the admins verify this patch?

nschonni commented on specification/web/resource-manager/Microsoft.Web/stable/2018-02-01/WebApps.json:
```diff
@@ -10252,36 +11106,29 @@
     },
-    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Web/sites/{name}/slots/{slot}/functions/{functionName}": {
-      "get": {
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Web/sites/{name}/slots/{slot}/host/default/listsyncstatus": {
```
```suggestion
    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Web/sites/{name}/slots/{slot}/host/default/listSyncStatus": {
```

mathewc commented on specification/web/resource-manager/Microsoft.Web/stable/2018-02-01/WebApps.json:
```diff
@@ -3334,44 +3347,69 @@
     },
-    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Web/sites/{name}/hybridConnectionNamespaces/{namespaceName}/relays/{relayName}": {
-      "get": {
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Web/sites/{name}/host/default/listsyncstatus": {
```
None of our other apis are doing camel casing like this, so I'd like to leave this consistent. If we were to make this change, I'd want to do it consistently across the board.

nschonni replied:
If you want to ignore it, you can added the word to https://github.com/Azure/azure-rest-api-specs/blob/master/custom-words.txt

mathewc replied:
Done, thanks.

mathewc replied:
@nschonni - are we good here now?

nschonni replied:
sorry, I'm not a real reviewer, feel free to resolve the conversations

mathewc replied:
I see, thanks. So what is the merge process for this repo? This will be my first commit here - am I waiting on @amarzavery ?

nschonni replied:
I think there is some workflow info in the [PR template](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/PULL_REQUEST_TEMPLATE.md), but some of it is internal to MS, so I'm not sure how it works

amarzavery commented:
@mathewc - Can you add x-ms-examples for all the new operations that have been added in the PR, along with modifying existing examples for the property `test_data_href` of model `FunctionEnvelope`

amarzavery commented:
Github is not showing the diff accurately. If you want to see the correct diff then you can see it on OpenAPIHub over [here](https://portal.azure-devex-tools.com/tools/diff/eyJzb3VyY2VVcmwiOiJodHRwczovL2dpdGh1Yi5jb20vQXp1cmUvYXp1cmUtcmVzdC1hcGktc3BlY3MvYmxvYi9tYXN0ZXIvc3BlY2lmaWNhdGlvbi93ZWIvcmVzb3VyY2UtbWFuYWdlci9NaWNyb3NvZnQuV2ViL3N0YWJsZS8yMDE4LTAyLTAxL1dlYkFwcHMuanNvbiIsInRhcmdldFVybCI6Imh0dHBzOi8vZ2l0aHViLmNvbS9BenVyZS9henVyZS1yZXN0LWFwaS1zcGVjcy9ibG9iLzc3ZGY0YmQxNDU3NTA3MDU0ZTA0YTE4ZTFlYzI4OWI4MTVkYWRiZmYvc3BlY2lmaWNhdGlvbi93ZWIvcmVzb3VyY2UtbWFuYWdlci9NaWNyb3NvZnQuV2ViL3N0YWJsZS8yMDE4LTAyLTAxL1dlYkFwcHMuanNvbiJ9).

amarzavery reviewed (changes requested):
Provide x-ms-examples for new operations and for updates made to existing model definitions.

mathewc commented:
Thanks. Most of our APIs currently don't have x-ms-examples, so these updates are consistent there. I'd prefer to handle adding of examples for all the APIs missing them as a separate work item, since it will be a bunch of work.

mathewc commented:
@amarzavery What do you think? If these changes look good themselves I say we get them in, and handle adding examples for all our APIs currently missing samples in a separate work item.

sanjaiganesh commented:
>             "invoke_url_template": {

how does user get list of templates ?

---
Refers to: specification/web/resource-manager/Microsoft.Web/stable/2018-02-01/WebApps.json:19720 in 77df4bd. [](commit_id = 77df4bd1457507054e04a18e1ec289b815dadbff, deletion_comment = False)

sanjaiganesh commented:
>             "language": {

enum ?

---
Refers to: specification/web/resource-manager/Microsoft.Web/stable/2018-02-01/WebApps.json:19724 in 77df4bd. [](commit_id = 77df4bd1457507054e04a18e1ec289b815dadbff, deletion_comment = False)

sanjaiganesh commented:
>             }

202 shouldn't return actual resource. 
202 requires returning azure-async header or location header

But  201 + provisioning State + azure async operation header pattern is recommended for async operations over the 202 + location header pattern. ref - [https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/Addendum.md#creating-or-updating-resources](https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/Addendum.md#creating-or-updating-resources)

---
Refers to: specification/web/resource-manager/Microsoft.Web/stable/2018-02-01/WebApps.json:3159 in 77df4bd. [](commit_id = 77df4bd1457507054e04a18e1ec289b815dadbff, deletion_comment = False)

sanjaiganesh commented:
>           "default": {

add examples.. also specify thos efiles using 'x-ms-examples":

---
Refers to: specification/web/resource-manager/Microsoft.Web/stable/2018-02-01/WebApps.json:3167 in 77df4bd. [](commit_id = 77df4bd1457507054e04a18e1ec289b815dadbff, deletion_comment = False)

sanjaiganesh commented:
>               "$ref": "#/definitions/KeyInfo"

202 shouldn't return actual resource. 
202 requires returning azure-async header or location header

But  201 + provisioning State + azure async operation header pattern is recommended for async operations over the 202 + location header pattern. ref - [https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/Addendum.md#creating-or-updating-resources](https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/Addendum.md#creating-or-updating-resources)

---
Refers to: specification/web/resource-manager/Microsoft.Web/stable/2018-02-01/WebApps.json:10875 in 77df4bd. [](commit_id = 77df4bd1457507054e04a18e1ec289b815dadbff, deletion_comment = False)

sanjaiganesh commented:
>           },

examples pls.. also specify those file paths using x-ms-examples

---
Refers to: specification/web/resource-manager/Microsoft.Web/stable/2018-02-01/WebApps.json:10883 in 77df4bd. [](commit_id = 77df4bd1457507054e04a18e1ec289b815dadbff, deletion_comment = False)

sanjaiganesh commented:
>     "FunctionEnvelope": {

it is not recommended to add properties in same api version. new api version required.
Could you explain/justify how you are going to handle clients calling with old and new sdks ?

---
Refers to: specification/web/resource-manager/Microsoft.Web/stable/2018-02-01/WebApps.json:19665 in 77df4bd. [](commit_id = 77df4bd1457507054e04a18e1ec289b815dadbff, deletion_comment = False)

sanjaiganesh commented:
>     "FunctionEnvelope": {

ex: client uses new sdks to specify these properties and then makes another call with old sdk, without these ? what happens to thse properties. 

---
In reply to: [508888205](https://github.com/Azure/azure-rest-api-specs/pull/6472#issuecomment-508888205) [](ancestors = 508888205)

---
Refers to: specification/web/resource-manager/Microsoft.Web/stable/2018-02-01/WebApps.json:19665 in 77df4bd. [](commit_id = 77df4bd1457507054e04a18e1ec289b815dadbff, deletion_comment = False)

sanjaiganesh commented:
>           "default": {

examples and list files using x-sms-examples

---
Refers to: specification/web/resource-manager/Microsoft.Web/stable/2018-02-01/WebApps.json:11100 in 77df4bd. [](commit_id = 77df4bd1457507054e04a18e1ec289b815dadbff, deletion_comment = False)

sanjaiganesh commented:
>         "description": "Get host secrets for a function app.",

update summary and description

---
Refers to: specification/web/resource-manager/Microsoft.Web/stable/2018-02-01/WebApps.json:11066 in 77df4bd. [](commit_id = 77df4bd1457507054e04a18e1ec289b815dadbff, deletion_comment = False)

sanjaiganesh commented:
>         "responses": {

200 ?


---
Refers to: specification/web/resource-manager/Microsoft.Web/stable/2018-02-01/WebApps.json:11142 in 77df4bd. [](commit_id = 77df4bd1457507054e04a18e1ec289b815dadbff, deletion_comment = False)

sanjaiganesh commented:
>         ],

examples.. ? x-ms-examples

---
Refers to: specification/web/resource-manager/Microsoft.Web/stable/2018-02-01/WebApps.json:11141 in 77df4bd. [](commit_id = 77df4bd1457507054e04a18e1ec289b815dadbff, deletion_comment = False)

sanjaiganesh commented:
>         ],

200 ? and examples x-ms-examples 


---
Refers to: specification/web/resource-manager/Microsoft.Web/stable/2018-02-01/WebApps.json:3389 in 77df4bd. [](commit_id = 77df4bd1457507054e04a18e1ec289b815dadbff, deletion_comment = False)

sanjaiganesh commented:
>         "responses": {

200 .. and x-ms-examples

---
Refers to: specification/web/resource-manager/Microsoft.Web/stable/2018-02-01/WebApps.json:3214 in 77df4bd. [](commit_id = 77df4bd1457507054e04a18e1ec289b815dadbff, deletion_comment = False)

sanjaiganesh commented:
I see lot of new apis around keys and added decent number of comments. please address the comments and schedule a meeting with arm api review board to discuss the need of these new apis + new properties being added to existing resources in same api version

sanjaiganesh commented:
>         "description": "Get host secrets for a function app.",

update descriptions here and other places

---
Refers to: specification/web/resource-manager/Microsoft.Web/stable/2018-02-01/WebApps.json:3328 in 77df4bd. [](commit_id = 77df4bd1457507054e04a18e1ec289b815dadbff, deletion_comment = False)

sanjaiganesh commented:
>           "404": {

it is actually 204 , if key not found
[[https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/resource-api-reference.md#delete-resource](https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/resource-api-reference.md#delete-resource)](https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/resource-api-reference.md#delete-resource)

The resource provider can return 200 (OK) or 204 (NoContent) to indicate that the operation completed successfully. A 200 (OK) should be returned if the object exists and was deleted successfully; and a 204 (NoContent) should be used if the resource does not exist and the request is well formed

---
Refers to: specification/web/resource-manager/Microsoft.Web/stable/2018-02-01/WebApps.json:3215 in 77df4bd. [](commit_id = 77df4bd1457507054e04a18e1ec289b815dadbff, deletion_comment = False)

sanjaiganesh commented:
>         ],

200 + examples x-ms-examples

---
Refers to: specification/web/resource-manager/Microsoft.Web/stable/2018-02-01/WebApps.json:3422 in 77df4bd. [](commit_id = 77df4bd1457507054e04a18e1ec289b815dadbff, deletion_comment = False)

sanjaiganesh commented:
>             }

pls refer to my comment in other places on 202

---
Refers to: specification/web/resource-manager/Microsoft.Web/stable/2018-02-01/WebApps.json:3484 in 77df4bd. [](commit_id = 77df4bd1457507054e04a18e1ec289b815dadbff, deletion_comment = False)

sanjaiganesh commented:
>         "responses": {

add 200
and itis 204 for key not found
[https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/resource-api-reference.md#delete-resource](https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/resource-api-reference.md#delete-resource)

---
Refers to: specification/web/resource-manager/Microsoft.Web/stable/2018-02-01/WebApps.json:3539 in 77df4bd. [](commit_id = 77df4bd1457507054e04a18e1ec289b815dadbff, deletion_comment = False)

sanjaiganesh commented:
>             "schema": {

move these to definitions and reference them in all these apis


---
Refers to: specification/web/resource-manager/Microsoft.Web/stable/2018-02-01/WebApps.json:10860 in 77df4bd. [](commit_id = 77df4bd1457507054e04a18e1ec289b815dadbff, deletion_comment = False)

sanjaiganesh commented:
>           "404": {

200 for delete success.. 204 for key not found (not 404.. remove)

also x-ms-examples

---
Refers to: specification/web/resource-manager/Microsoft.Web/stable/2018-02-01/WebApps.json:10939 in 77df4bd. [](commit_id = 77df4bd1457507054e04a18e1ec289b815dadbff, deletion_comment = False)

sanjaiganesh commented:
>         ],

200
x-,ms-examples

---
Refers to: specification/web/resource-manager/Microsoft.Web/stable/2018-02-01/WebApps.json:11181 in 77df4bd. [](commit_id = 77df4bd1457507054e04a18e1ec289b815dadbff, deletion_comment = False)

sanjaiganesh commented:
>               "$ref": "#/definitions/KeyInfo"

refer to my comment on 202 in other places.. + x-ms-examples

---
Refers to: specification/web/resource-manager/Microsoft.Web/stable/2018-02-01/WebApps.json:11249 in 77df4bd. [](commit_id = 77df4bd1457507054e04a18e1ec289b815dadbff, deletion_comment = False)

sanjaiganesh commented:
>           },

refer to my comment on delete response code in other places

---
Refers to: specification/web/resource-manager/Microsoft.Web/stable/2018-02-01/WebApps.json:11315 in 77df4bd. [](commit_id = 77df4bd1457507054e04a18e1ec289b815dadbff, deletion_comment = False)

mathewc commented:
Thanks @sanjaiganesh. The comments you made above don't take me to the line you’re commenting on when I click the links. Also, I can’t reply directly to them inline it seems. However, I can try line by line here:

1.	invoke_url_template – there is no “list of templates”. The template is simply the full http route to the Azure Function. It’s called a template, because if the route contains route params/expressions those will be returned, making the string a template
2.	language – language cannot be an enum because languages are extensible by 3rd parties in our system. There is no single complete list of them.
3.	202 status code – this was just an xml doc mistake. The API actually returns 201 Created. **I’ve fixed these 4 occurrences in the latest revision**
4.	x-ms-examples – there is an open tracking item https://github.com/Azure/azure-rest-api-specs/issues/6561 for adding examples for all the AppService APIs. None of our APIs have examples currently, and these changes are consistent with that. We’ll address that separately as part of a larger work item, to ensure consistency.
5.	New properties added to FunctionEnvelope – those new properties are GET only so are not a problem. They are not used in any PUT operation. These APIs are already live in production and are not causing any issues. 
6.	Summary/description comment – you mentioned “update summary and description”. From what I see all the summary/description comments for the host key APIs are correct. What problem are you referring to?
7.	404 key not found – suggestion was to return 204 rather than 404 for key not found. Looking across all our DELETE APIs, I see many that return 404s, and some that return 204s for not found. So we’re not consistent it seems. For these Azure Functions specific APIs, we return 404s when resources are not found. E.g. see existing API for DeleteFunction. Anyhow, this API is already live in production and being used, so we can’t change the status code at this point.
8.	Syncfunctiontriggers status code – your suggestion was to return 200 rather than 204? This new host/default/sync API returns the same response as the existing syncfunctionstriggers API which returns 204. So these must be consistent.
9.	“move these to definitions and reference them in all these apis” – I don’t understand the ask here. These schema blocks were auto-generated by our tools.
10.	“200 for delete success” – as noted above, these delete APIs are consistent with our other delete APIs like DeleteFunction, which in this swagger. The APIs are live and return 204 NoContent so this swagger is correct.

I’ve updated the PR fixing a few of the issues you mentioned. For all the others, I don’t think changes are needed based on my explanations. Thanks.

amarzavery commented:
@mathewc - When you say
> New properties added to FunctionEnvelope – those new properties are GET only so are not a problem. They are not used in any PUT operation. These APIs are already live in production and are not causing any issues.

`FunctionEnvelope` model is a request body parameter to `WebApps_CreateInstanceFunctionSlot` and `WebApps_CreateFunction` which are PUT operations. The new properties `"test_data_href", "invoke_url_template", "language", "isDisabled"` being added to `"FunctionEnvelope"` Model are not marked as `"readOnly": true` in the spec. Which means the user can send them in the request body of the above mentioned `PUT` operation.

Moreover the description of `"isDisabled"` says `"Gets or sets a value indicating whether the function is disabled"` which clearly indicates that not all property are `readOnly`. Thus this is indeed an API breaking change?

Why does your team not get the API changes reviewed in the design phase, before implementing these changes and deploying them to production?

mathewc commented:
While users may pass values for those, they are not used on PUT requests. Only the Name/Files/Config properties are used in the backend on Create operations. This existing API was created long ago by someone else - perhaps a different model type should have been used to make these points clear, however it was decided to reuse the same model type for GET/PUT. If users were to send these values they would be ignored by the backend.

Regarding the xml comments on these model types - these models map ultimately to the backend type https://github.com/Azure/azure-functions-host/blob/dev/src/WebJobs.Script.WebHost/Models/FunctionMetadataResponse.cs which this ARM API proxies requests to. When I added this new property, I just copied it including its xml comment. I didn't realize this comment would show up in our swagger. I've fixed the comment to clarify that the property is read only. Regarding why the various properties on this type were not marked "readonly" - I can't answer that because I didn't add the type initially. Perhaps @naveedaz can comment on that. Because my new additions are consistent with what is already in place I think this is fine. As an improvement I could add "readonly" to all these properties, but I'd rather not mess with the existing working properties. We haven't gotten any complaints on any of this.

These APIs were reviewed in the design phase before release, and are working fine in production, even after the model type additions shown in this PR. The "Gets or sets a value" doc issue is a minor thing that wasn't noticed in PR, and I've fixed that. At this point we're just trying to get our swagger updated to match production APIs so people can use the new APIs from the SDKs.

KrisBash commented:
@mathewc I wanted to provide some clarity on the delete responses as these are out of spec per [ARM RPC](https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/resource-api-reference.md#delete-resource): 
_The resource provider can return 200 (OK) or 204 (NoContent) to indicate that the operation completed successfully. A 200 (OK) should be returned if the object exists and was deleted successfully; and a 204 (NoContent) should be used if the resource does not exist and the request is well formed._

So, we'd expect to see 3 response definitions on any DELETE operation in swagger: 200, 204, and default. The RP should return 200 to indicate the resource was deleted in response, and 204 if the resource does not exist. 

Looking through the entirety of this swagger, some define 200+204, others 200+404, and others 204+404. 404's should be replaced with "default" with the standard error definition (which encompasses 404). Is the service this inconsistent in terms of 200 vs 204? We will want to see the service aligned to contract to ensure ARM and client expectations are met. Let me know about current service state vs what's in the swagger spec.

yungezz commented:
hi @mathewc , could you pls update PR to address comments? Or close the PR is the change is not ready for reviewing and merging.

yungezz commented:
@mathewc ping again.

mathewc commented:
@yungezz - this PR is being replaced by https://github.com/Azure/azure-rest-api-specs/pull/7174. Let's continue the review there.
